File: common-problems-leetcode/easy/remove-linked-list-elements.py
"""
Remove all elements from a linked list of integers that have value val.

Example:

Input:  1->2->6->3->4->5->6, val = 6
Output: 1->2->3->4->5
"""

# 64ms. 99th percentile.
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# 1->2->2->2->6->3->2
class Solution:
    def removeElements(self, head: ListNode, val: int) -> ListNode:
        dummy = ListNode(-1)
        dummy.next = head
        runner = dummy
        
        while runner:
            while runner.next:
                if runner.next.val == val:
                    runner.next = runner.next.next
                else:
                    break
            
            runner = runner.next
        
        return dummy.next

File: common-problems-leetcode/easy/remove-linked-list-elements.java
/*
Remove all elements from a linked list of integers that have value val.

Example:

Input:  1->2->6->3->4->5->6, val = 6
Output: 1->2->3->4->5
*/
// 1ms. 98th percentile. 


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode removeElements(ListNode head, int val) {
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode runner = dummy;
        
        while (runner != null) {
            while (runner.next != null) {
                if (runner.next.val == val) {
                    runner.next = runner.next.next;
                } else {
                    break;
                }
            }
            runner = runner.next;
        }
        return dummy.next;
    }
}